### PR TITLE
Refactor: change signature of format()

### PR DESCRIPTION
**Pull request by monitotxi**
Change signature of format() method to accept CurrencyFormat interface.

This is the first step to be able to improve the format() method as requested on #1 

This is a breaking change.

### CODE REVIEW

**Comment by castarco** [comment]
Hi,

I'm not convinced about the proposed interface. I'll post here a list of desirable properties that should be met:

* [ ] Allow choosing between printing the symbol or the ISO code.
* [ ] Specify if we want a space between the symbol/ISO_Code and the monetary amount.
* [ ] Allow choosing (in `CurrencyFormat` construction time) between specifying the desired precision... OR the *extra* precision beyond the official currency's precision.
* [ ] **composability** : `CurrencyFormat` instances should be able to hold shareable formatting data between currencies, and combine its formatting data with the formatting data provided by the `CurrencyType` instance.

I think that, whatever we do, we'll have to add backwards breaking changes, so it's interesting to discuss it in detail to avoid having to break compatibility two times instead of one single time.

**Comment by monitotxi** [comment]
Hi,

I agree with the list posted by @castarco. I've added 2 new methods to the CurrencyFormat interface and adapted the format method to use them.

But I'm not sure I understand the *composability* property. Can you provide an example?

File: src/Currency.php [hunk 1]
```diff
@@ -144,16 +145,31 @@ public function getAmountAsFractionalUnits(): int
...
-        $nDecimals = $this->currencyType->getNumFractionalDigits() + $extraPrecision;
+        if (is_null($currencyFormat)) {
+            $currencyFormat = new CurrencyFormat();
+        }
+
+        $precision = $currencyFormat->getPrecision();
+        if (is_null($precision)) {
+            $precision = $this->currencyType->getNumFractionalDigits();
+        }
+        $nDecimals = $precision + $currencyFormat->getExtraPrecision();
```
Comment (castarco) [inline src/Currency.php]: This is one example of the composability that i referred in previous comments:

1. If the `$currencyFormat` instance returns a not-null value when we call `getPrecision`, the we should use this value.
2. If `getPrecision` returns null, then we should take the currency type precision.
3. If the `getExtraPrecision` method of the `$currencyFormat` instance returns a not-null value (AND the `$currencyFormat->getPrecision` call returned null), then, sum it to the currency type default precision.

The lines 153-161 don't match this expected behaviour.

Comment (castarco) [inline src/Currency.php]: The idea behind this "strange" desired behavior is to allow complete flexibility to specify how many digits we want to handle. We can specify the absolute number of digits, or an extra number of digits over the "default" digits specified by the currency type.

Comment (monitotxi) [inline src/Currency.php]: Fixed at 8f1211b2d531288013d681cf970c900f79781867

File: src/CurrencyFormat.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * @var int
+     */
+    private $precision;
+
+    public function __construct(
+        string $decimalsSeparator = '.',
+        string $thousandsSeparator = '',
+        int $extraPrecision = 0,
+        int $decorationType = self::DECORATION_NO_DECORATION,
```
Comment (castarco) [inline src/CurrencyFormat.php]: This default does not match the current behavior (see the last three lines of the `Currency::format` method)

File: src/Currency.php [hunk 3]
```diff
@@ -144,16 +145,30 @@ public function getAmountAsFractionalUnits(): int
...
+        $number = ('' === $currencyFormat->getThousandsSeparator())
+            ? \str_replace('.', $currencyFormat->getDecimalsSeparator(), $amount->__toString())  // This is safer!
+            : \number_format(
+                $amount->asFloat(),
+                $nDecimals,
+                $currencyFormat->getDecimalsSeparator(),
+                $currencyFormat->getThousandsSeparator()
+            );
 
         return ($this->currencyType->getSymbolPlacement() === CurrencyType::BEFORE_PLACEMENT)
```
Comment (castarco) [inline src/Currency.php]: Here we can use the `$currencyFormat->getDecorationType` call to know if we have to use the currency symbol, the currency ISO code, or no decoration at all.

Remember that the "symbol placement" is only valid for symbols, ISO codes are always placed after the numeric amount.

File: src/CurrencyFormat.php (hunk 2 above)
Comment (monitotxi) [inline src/CurrencyFormat.php]: Modified to match the default behaviour.

File: src/Currency.php (hunk 3 above)
Comment (monitotxi) [inline src/Currency.php]: Agree. It was in a TODO comment, but I just added it.

**Pull request merged**